src/aggregator/rate/limiter.go

```diff
@@ -21,77 +21,74 @@
 package rate
 
 import (
-	"sync"
-	"sync/atomic"
 	"time"
 
-	"github.com/m3db/m3/src/x/clock"
+	"go.uber.org/atomic"
+	"golang.org/x/sys/cpu"
+
+	xtime "github.com/m3db/m3/src/x/time"
 )
 
 var (
-	zeroTime = time.Unix(0, 0)
+	zeroTime = xtime.UnixNano(0)
 )
 
 // Limiter is a simple rate limiter to control how frequently events are allowed to happen.
+// It trades some accuracy on resets for speed.
 type Limiter struct {
-	sync.RWMutex
-
-	limitPerSecond int64
-	nowFn          clock.NowFn
-
-	alignedLast time.Time
-	allowed     int64
+	limitPerSecond atomic.Int64
+	alignedLast    atomic.Int64
+	_              cpu.CacheLinePad // prevent false sharing
+	allowed        atomic.Int64
 }
 
 // NewLimiter creates a new rate limiter.
-func NewLimiter(l int64, fn clock.NowFn) *Limiter {
-	return &Limiter{
-		limitPerSecond: l,
-		nowFn:          fn,
-	}
+func NewLimiter(l int64) *Limiter {
+	limiter := &Limiter{}
+	limiter.limitPerSecond.Store(l)
+	return limiter
 }
 
-// Limit returns the current limit.
+// Limit returns the current limit. A zero limit means no limit is set.
 func (l *Limiter) Limit() int64 {
-	l.RLock()
-	limit := l.limitPerSecond
-	l.RUnlock()
-	return limit
+	return l.limitPerSecond.Load()
 }
 
 // IsAllowed returns whether n events may happen now.
 // NB(xichen): If a large request comes in, this could potentially block following
 // requests in the same second from going through. This is a non-issue if the limit
 // is much bigger than the typical batch size, which is indeed the case in the aggregation
 // layer as each batch size is usually fairly small.
-func (l *Limiter) IsAllowed(n int64) bool {
-	alignedNow := l.nowFn().Truncate(time.Second)
-	l.RLock()
-	if !alignedNow.After(l.alignedLast) {
-		isAllowed := atomic.AddInt64(&l.allowed, n) <= l.limitPerSecond
-		l.RUnlock()
-		return isAllowed
+// The limiter is racy on window changes, and can be overly aggressive rejecting requests.
+// As the limits are usually at least 10k+, the error is worth the speedup.
+func (l *Limiter) IsAllowed(n int64, now xtime.UnixNano) bool {
```

Comment:
aside, `IsAllowed` probably isn't the right verbiage here (i know you didn't write it, but just pointing it out) - "is"-style names sound like accessors or readonly evaluations, whereas an "Add" or something would make it clear that the call may mutate internal state.

not asking you to change it here, might be a rainy day thing we clean up later when we're doing other hygiene things.